lucille-core/src/main/java/com/kmwllc/lucille/indexer/OpenSearchIndexer.java

```diff
@@ -96,9 +104,107 @@ protected void sendToIndex(List<Document> documents) throws Exception {
       return;
     }
 
-    BulkRequest.Builder br = new BulkRequest.Builder();
+    Map<String, Document> documentsToUpload = new LinkedHashMap<>();
+    Set<String> idsToDelete = new LinkedHashSet<>();
+    Map<String, List<String>> termsToDeleteByQuery = new LinkedHashMap<>();
 
+    // populate which collection each document belongs to
+    // if document is marked for deletion ONLY, then only add to idsToDelete
+    // if document is marked for deletion AND contains deleteByFieldField and deleteByFieldValue, only add to termsToDeleteByQuery
+    // else then add to upload
     for (Document doc : documents) {
+      String id = doc.getId();
+      if (isMarkedForDeletion(doc)) {
+        documentsToUpload.remove(id);
+        if (isMarkedForDeletionByField(doc)) {
+          String field = doc.getString(deleteByFieldField);
+          if (!termsToDeleteByQuery.containsKey(field)) {
+            termsToDeleteByQuery.put(field, new ArrayList<>());
+          }
+          termsToDeleteByQuery.get(field).add(doc.getString(deleteByFieldValue));
+        } else {
+          idsToDelete.add(id);
+        }
+      } else {
+        idsToDelete.remove(id);
+        documentsToUpload.put(id, doc);
+      }
+    }
+
+
+    uploadDocuments(new ArrayList<>(documentsToUpload.values()));
+    deleteById(new ArrayList<>(idsToDelete));
+    deleteByQuery(termsToDeleteByQuery);
+  }
+
+  private void deleteById(List<String> idsToDelete) throws Exception {
+    if (idsToDelete.isEmpty()) {
+      return;
+    }
+
+    BulkRequest.Builder br = new BulkRequest.Builder();
+    for (String id : idsToDelete) {
+      br.operations(op -> op
+          .delete(d -> d
+              .index(index)
+              .id(id)
+          )
+      );
+    }
+
+    BulkResponse response =  client.bulk(br.build());
+    if (response.errors()) {
+      for (BulkResponseItem item : response.items()) {
+        if (item.error() != null) {
+          log.debug("Error while deleting id: {} because: {}", item.id(), item.error().reason());
+        }
+      }
+      throw new IndexerException("encountered errors while deleting documents");
+    }
+  }
+
+  private void deleteByQuery(Map<String, List<String>> termsToDeleteByQuery) throws Exception {
+    if (termsToDeleteByQuery.isEmpty()) {
+      return;
+    }
+    // termsToDeleteByQuery is never empty
+    BoolQuery.Builder boolQuery = new BoolQuery.Builder();
+    for (Map.Entry<String, List<String>> entry : termsToDeleteByQuery.entrySet()) {
+      String field = entry.getKey();
+      List<String> values = entry.getValue();
+      boolQuery.should(f -> f
+        .terms(t -> t
+          .field(field)
+          .terms(tt -> tt.value(values.stream()
+              .map(FieldValue::of)
+              .collect(Collectors.toList()))
```

Comment:
This boolQuery is represented as such below. Each entry would be adding a terms query to the should clause and each of the terms query would contain a field (entryKey) and fieldValue(s) (entryValue). Lastly, we set minimumShouldMatch to 1, such that any document that passes any of the terms query would show up in resultSet. For a terms query to pass, a document would need to contain the field and in that field contain any one of the field values.
```
{
  "query": {
    "bool": {
      "should": [
        {
          "terms": {
            "field1": ["value1", "value2", "value3"]
          }
        },
        {
          "terms": {
            "field2": ["valueA", "valueB"]
          }
        }
      ],
      "minimum_should_match": 1
    }
  }
}
```

Reply:
I think it would be valuable to include this as a comment in the code itself

Reply:
added comment in [af56102](https://github.com/kmwtechnology/lucille/pull/191/commits/af56102e1864b5297de99dfa81a184026483d666)